benchopt/base.py

```diff
@@ -87,7 +103,7 @@ def set_objective(self, **objective_dict):
         """
         ...
 
-    @abstractmethod
+    @not_implemented
```

Comment:
this was correct. It's an ABC class. abstractmethod in an ABC class are the methods to be implemented by subclasses. It's checked at runtime.

Reply:
The abstract class does not allow you to implement one of 2 methods. Here we want to have either `run` or `run_with_cb` implemented. so `not_implemented` is a mechanism to allow to check this while still having API definition in the class.

Reply:
how about we enforce that run is always available but run_with_cb is optional. If it's available it's preferred if option says so. This would avoid duplicating all solvers and to use the same solver code with an without tracing using a callback?

Reply:
I don't understand why we would need to duplicate the solver?
The solver should just implement one of the 2, not both. Then we can either call with a dummy cb or the cb that log everything.  I don't think requiring to implement both would be super user friendly.